### DISCUSSION TITLE
Hook Developer Request

### DISCUSSION THREAD

**Discussion by MarvinSunday** [XRPLF/XRPL-Standards#181] [Looking for help]
Title: Hooks Developer for Our Marketplace (Foodware Hub)
Revision: 1 (13-02-2024)
### 
Author: [Marvin Sunday](https://linktr.ee/marvinsunday/)
Affiliation: [Foodware Logistics Technologies](https://linktr.ee/foodware_xrpl/)
### Abstract
Over the ages, humans have continued to seek a means to transact value through barter, tools, metals & minerals, Coins, IOUs, Fiat, and now Crypto. 
Food is one utility essential to man's survival on Earth, a gift of nature. Sadly, it is not so in Africa today; Food has become scarce despite being rich with fertile land for agriculture and other allied activities.
We envision a system that recycles value between farmers, food vendors, and consumers and ensures everyone has the value they deserve.
We intend to do this with Blockchain through the XRP Ledger protocol and Hooks on the Xahau.
### Introduction
Foodware Logistics Technologies Ltd. is a Nigerian-based Blockchain-Powered Food distribution company (Start-up) created from the Hult Prize Competition in the year 2020. Its Mission is to reengineer the Food distribution cycle in Africa to provide more value for its stakeholders (Farmers, Vendors, and Consumers) see [Food Paper 1.0](https://drive.google.com/file/d/1WibTk0rVVxuVEN3K8UCZif5itXIt7bfn/view?usp=sharing).
Foodware is rapidly evolving to integrate efficient and futuristic technologies into its infrastructure.
Blockchain is an integral part of its value/monetary transactions system, and it intends to do so in a zero-knowledge process with the help of Xahau's Hooks software.
This Discussion is to list its intended use of the technology and to call in Hook Developers to its project.
### Description
We are building a Decentralized Food marketplace and Delivery Affiliate systems that will evolve into a Food Super App. The Xahau Ledger is our intended value layer technology as we envision the ability to turn XRPL wallets into User accounts with XRPL Hooks software.
We need Hooks to build the Following;
- A Guard Hook for all our User Accounts
- An Escrow Hook for purchase and Delivery timing and execution
- A Crowdfund & Subsidy Hook for collective purchases
- A Reward Hook for incentivizing our users
- A Voting/polling Hook for consumer Governance
- A Discount Hook for our Cashback system.

 These Hooks will interact with the Following Ledgers objects.
- Hooks (themselves)
- Escrow Objects (XAH & Tokens)
- Fungible Tokens (Stablecoins, L1s, and Xahau Tokens)
- URITokens
- Destination Tagged XRPL Accounts (X-wallets)
- Remit Transactions
- Invoke transactions from Hotpocket dApps for Turing complete automation.
### Requirements for Hook Developer
- The individual should have a strong background and Experience in C, C++, and the XRP Ledger code infrastructure.
- The individual should have prior understanding and Familiarity with Hooks and Hooks Building.
- The Individual should be familiar with Evernodes and Hotpocket dApps or know how to build with them. Both would be highly preferable.
- The individual should be able to integrate these infrastructures into web Apps using the Available SDKs and APIs on the XRP Ledger.
- Lastly, we want a thinker or thinkers as we don't consider our research results absolute. We welcome improvement as long as it is within the scope of our plans.
### Conclusion
We are open to discussions at this point.
Please cite your interests or concerns here in the comment sessions.
You can access more of our resources [here](https://linktr.ee/foodware_xrpl/)
You can email us here at foodwarexrpl@gmail.com

**Comment by MarvinSunday** [comment]


**Reply by WietseWind** [reply]
Hey @MarvinSunday - thanks for posting! Makes it easy to share and follow up with questions / ...

Are you familiar with MermaidJS? It could (would) help to outline the suggested/proposed/ideal flow for the required Hooks. A short TL;DR (what would they do, use case) and what would be the flow. 

Also: is there a V1/V2? Looking at your post, I suppose this is what it would look like if everything gets realised, is there a V1 with must haves you can more easily accomplish? It could help breaking it up. Finding a dev for all of it could be harder, while someone could consider taking the plunge as a personal challenge if it's a much more comprehensible endeavour (thanks to simpler & clearer definitions of requirements & flow).

-- FYI: moved from `rippled` to `Standards Proposals` section. Also not really a standard proposal, but possibly a better place for this thread than the `rippled` node source code repo :)

**Reply by MarvinSunday** [reply]
Okay thanks Wieste

**Reply by MarvinSunday** [reply]
First off 
Thank you very much Wieste for showing up...
I'm not a developer and I just happened to understand how these codes work and I like working with developers to bring my ideas to life. I'm not familiar with mermaid JS but I have a flow document that will explain the flow of transactions on the hook but I think it's IP and I'm quite worried sharing it in public. Is there something we can do about this?
Yes there is a version 1, 2, 3, 4, ...
For example the first version will require just three hooks 
The Guard Hook for users.
The Escrow Hook for Delivery timing
Specialized Guard Hook for our X-wallets  
We would also have to host a small Hotpocket dApp for delivery timing automation.
Lastly, a simple XAH native Dex UI for Stablecoin to Foodware Coin trading.
I'll add it to my gist on GitHub and I'll drop the link here for easy access or should we leave it here since it has been moved? @WietseWind

**Reply by WietseWind** [reply]
@MarvinSunday Then I can recomment Whimsical, or any other flowchart app for that matter.

I understand how some of these things can be sensitive to share. This is where I encourage you to reconsider (no pressure, obviously, and there can be very good reasons not to share publicly).

But: so many people have great ideas, but because they don't share them, they potentially miss out on people who would want to collaborate. Opposed to the low risk that someone runs with it (as they don't have the background info, vision, etc.). But: no pressure. In my experience, out of all ideas shared, none of it has really been 'stolen'. While it did result in working together with the wonderful team we have right now at XRPL Labs. Also: just a bit more information about version one of certain Hooks would allow me to ask people in my team to take a look, and possibly (if not too much) dedicate some time to it to help you (project) on your way.

But: 100% respect if you prefer not to, and prefer to keep that for whenever you find someone interested in building alongside of you. I'll happily share this post here and there.

Re: moving to Gist: that works, but having moved this from the `rippled` repo to this repo, I think we're good for now :)

**Reply by MarvinSunday** [reply]
Reading what you've said I feel relieved and I'll come up with the necessary documents need for the first phase, and I believe in sharing and helping. I also understand that you want to help so I'll be honest while doing so...
I'll revert back to you in the morning after I get the document fixed for ths first phase ready.
Thanks a million Wieste!
@WietseWind

**Comment by MarvinSunday** [comment]
Good Afternoon Wieste I'm sorry for the Delay but I was trying to get some stuffs fixed while preparing the Documents 
So I was able to get a rough flowchart for the 3 Initial Hooks.
I'm attaching that document herewith and I hope you take a look and tell me what you think.
Thanks @WietseWind 
[Foodware Hooks V1.pdf](https://github.com/XRPLF/XRPL-Standards/files/14298904/Foodware.Hooks.V1.pdf)

**Reply by WietseWind** [reply]
Thanks, this should help those in a position to help with coding to decide if they are up for the job (knowledge and resource wise).

Three two questions:

1. The scope: is my understanding correct that these hooks are put in place to build a "marketplace" where (wholesale?) food suppliers (supply) and demand meet, where traceability & escrow-like payments are made possible?

2. Is my understanding correct that the Hooks are basically basically "logic based routers", deciding which amounts of funds go where, at a given time/event?

**Reply by MarvinSunday** [reply]
I appreciate your response Wieste, thanks for replying...
1. They are Made to build a marketplace and the first Hook controls user transaction inflow/outflow. The second and third Hook allows us to add delivery timeframe in Ledgers and conditioned escrow-like payments to the system, like you said. Yes, both wholesale and retail food suppliers can be present to meet user demands.
2. Yes, they are logic based routers that route funds based on the instructions from the Prompter Cluster (Hotpocket dApp) and the Foodware Accounts Wallet. Both the Prompter and Accounts wallet have multi-sign enabled to optimize security, the first uses Evernode.
If you need anymore clarity please lemme know @WietseWind

**Reply by WietseWind** [reply]
Thanks, that's it for now. This is good info for those interested in building. If you send out another Tweet (linking here, calling on XRPL Devs) I'm happy to RT. If we have time available ourselves, I know where to find you.

In the mean time: three things possibly deserving some attention (well meant feedback):

1. In your Food Paper 1.0, the problem statement and the solution don't seem to address the same thing?
2. Try to find interested parties for a pilot, and make sure they can operate in an environment that works: I just saw news that more African countries blocked exchange access, for this to work, local currencies must be available on ledger, or at least available for exchange somewhere to local currency: I've seen many great ideas end at the PoC stage because onramp/offramp (or lack of » pilot users) were a problem. Try to build a viable solution (more fun, real life use) over a technical possible solution)
3. Define user stories, and flows from the user story perspective. Try to simplify your flows. I have a feeling some of the Hook flows can be simplified ;)

**Reply by MarvinSunday** [reply]
1. That's I'll agree is true, we wrote the whitepaper a couple of years back precisely 2021, we haven't really gotten what we wanted to based on at that moment. We do now and we are changing the Document to suit that thanks for the heads up.
2. This is true and I agree, we think Nigeria will not ban exchanges due to the high source of revenue gotten from the Crypto by Nigerian citizens the Central Bank of Nigeria have recently lifted the Ban on Crypto to Fiat on/off ramp via banks. nevertheless, we know that the Onramp and offramp is an issue ideas of our kind face, so we are following closely an organisation spearheaded by Nigerian Banks and Fintechs "The African Stablecoin Consortium" as they wish to issue a regulatory and complaintly backed Naira Stablecoin called "Consortium Naira" (cNGN). We hope that this will make things work but considering your advice, a pilot in another area will be added to our plans.
3. I'll give it my best shot but like I said in the introduction I'm not a developer 😂😂😂, so it's only my perspective of how the hooks can be built and I'm open to modifications and changes. I'm also vividly trying to understand the users more to make these changes.
Thanks a Million @WietseWind

[a comment by an automated account is not shown]
